File: data structures and algorithms/python/06_076.py
def quickselect(arr, k):
    def partition(left, right, pivot_index):
        pivot_value = arr[pivot_index]
        # Move pivot to the end
        arr[pivot_index], arr[right] = arr[right], arr[pivot_index]
        
        store_index = left
        for i in range(left, right):
            if arr[i] < pivot_value:
                arr[store_index], arr[i] = arr[i], arr[store_index]
                store_index += 1
        
        # Move pivot to its final place
        arr[right], arr[store_index] = arr[store_index], arr[right]
        return store_index
    
    def select(left, right, k):
        if left == right:
            return arr[left]
        
        # Choose pivot (using middle element for simplicity)
        pivot_index = (left + right) // 2
        pivot_index = partition(left, right, pivot_index)
        
        if k == pivot_index:
            return arr[k]
        elif k < pivot_index:
            return select(left, pivot_index - 1, k)
        else:
            return select(pivot_index + 1, right, k)
    
    return select(0, len(arr) - 1, k)

# Example usage with detailed steps
def demo_quickselect():
    arr = [3, 2, 1, 5, 6, 4]
    k = 2  # Find 3rd smallest element (0-based index)
    
    print("Original array:", arr)
    result = quickselect(arr, k)
    print(f"The {k+1}th smallest element is: {result}")
    print("Note: Array may be partially sorted after running quickselect:", arr)

if __name__ == "__main__":
    demo_quickselect()

File: data structures and algorithms/cs/06_076.cs
using System;

public class QuickSelect
{
   /**
    * Finds the kth smallest element in an unordered array
    * @param arr Input array of integers
    * @param k The k-th smallest element to find (0-based)
    * @return The k-th smallest element in the array
    */
   public static int Quickselect(int[] arr, int k)
   {
      return Select(arr, 0, arr.Length - 1, k);
   }

   private static int Partition(
      int[] arr,
      int left,
      int right,
      int pivotIndex
   )
   {
      int pivotValue = arr[pivotIndex];
      Swap(arr, pivotIndex, right);

      int storeIndex = left;
      for (int i = left; i < right; i++)
      {
         if (arr[i] < pivotValue)
         {
            Swap(arr, storeIndex, i);
            storeIndex++;
         }
      }

      Swap(arr, right, storeIndex);
      return storeIndex;
   }

   private static int Select(
      int[] arr,
      int left,
      int right,
      int k
   )
   {
      if (left == right)
      {
         return arr[left];
      }

      int pivotIndex = (left + right) / 2;
      pivotIndex = Partition(arr, left, right, pivotIndex);

      if (k == pivotIndex)
      {
         return arr[k];
      }
      else if (k < pivotIndex)
      {
         return Select(arr, left, pivotIndex - 1, k);
      }
      else
      {
         return Select(arr, pivotIndex + 1, right, k);
      }
   }

   private static void Swap(int[] arr, int i, int j)
   {
      int temp = arr[i];
      arr[i] = arr[j];
      arr[j] = temp;
   }

   public static void Main(string[] args)
   {
      int[] arr = {
         3, 2, 1,
         5, 6, 4
      };
      int k = 2; // Find 3rd smallest element (0-based)

      Console.WriteLine(
         "Original array: " + ArrayToString(arr)
      );
      int result = Quickselect(arr, k);
      Console.WriteLine(
         $"The {k + 1}th smallest element is: {result}"
      );
      Console.WriteLine(
         "Note: Array may be partially sorted " +
         "after running quickselect: " +
         ArrayToString(arr)
      );
   }

   private static string ArrayToString(int[] arr)
   {
      return "[" + string.Join(", ", arr) + "]";
   }
}